lib/prefix.h

```diff
@@ -151,7 +151,7 @@ struct flowspec_prefix {
 /* FRR generic prefix structure. */
 struct prefix {
 	uint8_t family;
-	uint8_t prefixlen;
+	uint16_t prefixlen;
```

Comment:
prefix.h has bunch of prefix_length variables declared as uint8_t. There are 4 such instances. For example one in struct evpn_macip_addr. In various parts of BGP code, prefixlen (16 bit) is assigned to prefix_length (8 bit) variable.

I think these 4 needs to be changed to uint16_t.

Reply:
prefix_length is there in few more places in the code (not just BGP) and it was defined as uint8_t.

Reply:
Also, in few places stream_getc and stream_putc functions are used to copy/read data from/to prefixlen. stream_getc and stream_putc operates on takes/return uint8_t. It is good to review this usage along with this change.

Reply:
+1 to that, there are several more places that need updating for this.

Reply:
@srimohans 
1. The 4 places you referenced are fields holding the prefix length for an IP address, not fields indicating the overall length of the abstract `prefix` stored in `struct prefix`; I believe those do not need to be updated.
2. Regarding the assignment from a 16-bit to an 8-bit as performed in bgpd, we have:
```
* dst prefix must be AF_INET or AF_INET6 prefix, to forge EVPN prefix */
extern int bgp_build_evpn_prefix(int evpn_type, uint32_t eth_tag,
                                 struct prefix *dst)
{
...
        struct prefix p2;
        struct prefix *src = &p2;
...
        /* store initial prefix in src */
        prefix_copy(src, dst);
...
        p_evpn_p->route_type = evpn_type;
...
                p_evpn_p->prefix_addr.ip_prefix_length = p2.prefixlen;
```
Since p2.prefixlen should (per the function comment) be AF_INET or AF_INET6 this assignment should be valid.

Similarly in the other relevant place:
```
static inline void build_type5_prefix_from_ip_prefix(struct prefix_evpn *evp,
                                                     struct prefix *ip_prefix)
{
...
        evp->prefix.prefix_addr.ip_prefix_length = ip_prefix->prefixlen;
```
Again the assignment should be OK here.

3. Regarding the `stream_putc*` and `stream_getc*` question - `stream_putc` correctly takes a `uint8_t`, and the only places it is given `prefixlen` are places where the value of `prefixlen` ought to fit in 8 bytes. Similarly assignments from `stream_getc` into `prefixlen` will always work since that is a safe promotion from `uint8_t` to `uint16_t`; I tried to do a bit of additional validation and look for where the values that go into a `prefixlen` from a `stream_getc` come from and did not find any that should need more than 8 bits.

Reply:
@qlyoung at the very least this needs to be in a comment in the code because someone will definitely shoot themselves in the foot with this at some point in the future because they aren't aware prefixes are passed around with an 8-bit prefixlen.

Also, we could just do the work now and pass it around with a 16-bit prefixlen value, it's not exactly going to cause harm...